### PR TITLE
Leaderboards - StudentScore

### PR DESCRIPTION
**Pull request by sjagla27**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context
To make the leaderboards for my bachelor thesis more efficient we want to have the scores of the students closer with the exercises themselves.
Also see Issue #2044.

### Description
There will be two new tables added to the database. The student_score table will include the most recent result and score of an exercise for a student.

### Steps for Testing
best tested locally because currently there are no changes that can be seen or are sent to the client
1. Create exercises and participate in them with different users.
2. Assess the submissions.
3. Have a look in your db and check if all StudentScores are entered correctly.
4. Change assessments.
5. Step 3.

### Test Coverage
- StudentScore: 100%
- StudentScoreResource: 100%
- StudentScoreService: 93%

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/Result.java [hunk 1]
```diff
@@ -107,6 +110,12 @@
     @Column(name = "example_result")
     private Boolean exampleResult;
 
+    @OneToOne(mappedBy = "result", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/Result.java:113]: Why lazy? Should not be a huge overhead to have it eager imo as it is OneToOne :)

File: src/main/java/de/tum/in/www1/artemis/domain/scores/StudentScore.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+public class StudentScore extends DomainObject {
+
+    @ManyToOne
+    private User student;
+
+    @ManyToOne
+    private Exercise exercise;
+
+    /**
+     * A submission can have a result and therefore, results are persisted and removed with a submission.
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/domain/scores/StudentScore.java:21]: I do not understand this comment at this position?

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java [hunk 3]
```diff
@@ -110,6 +113,9 @@ public ProgrammingExerciseGradingService(ProgrammingExerciseTestCaseService test
             result = resultRepository.save(result);
             result.setSubmission(tmpSubmission);
 
+            // update StudentScore
+            studentScoreService.updateResult(result);
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java:117]: I am not an expert with entity listeners. Why is this needed when you have the entity listener?

File: src/main/java/de/tum/in/www1/artemis/service/StudentScoreService.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    /**
+     * Returns all StudentScores for course.
+     *
+     * @param course course
+     * @return list of student score objects for that course
+     */
+    public List<StudentScore> getStudentScoresForCourse(Course course) {
+        return studentScoreRepository.findAllByExerciseIdIn(course.getExercises());
```
Comment (stefanwaldhauser) [inline src/main/java/de/tum/in/www1/artemis/service/StudentScoreService.java:50]: Is there a nicer way to maybe do this in the query. I am not an expert but should be possible, no?

**Review by stefanwaldhauser — APPROVED** [review]
Had not time to test it yet but noticed some things that are easy to fix in the code :) GJ

File: src/main/java/de/tum/in/www1/artemis/domain/scores/StudentScore.java (hunk 2 above)
Comment (sjagla27) [reply src/main/java/de/tum/in/www1/artemis/domain/scores/StudentScore.java:21]: True. No idea. Removed.

File: src/main/java/de/tum/in/www1/artemis/service/StudentScoreService.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            studentScoreRepository.delete(score);
+        }
+    }
+
+    /**
+     * Returns one StudentScores for exercise and student if one exists.
+     *
+     * @param student student user
+     * @param exercise exercise
+     * @return list of student score objects for that course
```
[the remark this answers is not in the document]
Comment (sjagla27) [reply src/main/java/de/tum/in/www1/artemis/service/StudentScoreService.java:71]: Adressed.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java (hunk 3 above)
Comment (sjagla27) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java:117]: This is a special case that needs to be done separately.

File: src/main/java/de/tum/in/www1/artemis/service/StudentScoreService.java (hunk 4 above)
Comment (sjagla27) [reply src/main/java/de/tum/in/www1/artemis/service/StudentScoreService.java:50]: I don't really know what you mean. There will be additions to the queries when the client stuff gets implemented.

File: src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        // Remove StudentScore
+        studentScoreService.getObject().removeResult(deletedResult);
+    }
+
+    /**
+     * Before result gets updated, update all Scores.
+     *
+     * @param updatedResult updated result
+     */
+    @PreUpdate
```
[the remark this answers is not in the document]
Comment (sjagla27) [reply src/main/java/de/tum/in/www1/artemis/service/listeners/ResultListener.java:49]: The update of existing scores does not work with PostUpdate. It does however work with PreUpdate.

**Review by LudwigWeide — APPROVED** [review]
Tested on TS2. Works as intended

**Comment by philgabr** [comment]
Seems to work as intended.
However, I can not access the DB (tested on ts2), so I could check step (3) only on Artemis...

File: src/main/java/de/tum/in/www1/artemis/service/StudentScoreService.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    /**
+     * Updates all StudentScores for result updatedResult.
+     *
+     * @param updatedResult result to be updated
+     */
+    public void updateResult(Result updatedResult) {
+        // ignore results without score or participation
+        if (updatedResult.getScore() == null || updatedResult.getParticipation() == null || !Boolean.TRUE.equals(updatedResult.isRated())) {
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/StudentScoreService.java:94]: Just as a side note but I don't really understand why we use `!Boolean.TRUE.equals`. Wouldn't `Boolean.FALSE.equals` we easier to read?

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentScoreResource.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+/**
+ * REST controller for managing Rating.
+ */
+@Validated
+@RestController
+@RequestMapping("/api")
+public class StudentScoreResource {
+
+    // private static final String ENTITY_NAME = "studentScores";
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentScoreResource.java:36]: I think you forgot to remove this

**Review by ivanchimeno — COMMENTED** [review]
Didn't test yet but they code looks good

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentScoreResource.java (hunk 8 above)
Comment (sjagla27) [reply src/main/java/de/tum/in/www1/artemis/web/rest/StudentScoreResource.java:36]: True.

**Review by krusche — CHANGES_REQUESTED** [review]
we would need server integration tests for all cases in which results are created or changed, e.g.:
1. Programming Exercise --> new automatic result, manual results (save, submit, override, complaints)
2. Programming Grading Service --> edit result (re-evaluate)
3. Quiz Exercises --> create, re-evaluate
4. Modeling/Text/File upload: Save assessment, submit assessment, override assessment, complaints (edit result)

**Comment by krusche** [comment]
I close this PR for now as it is open way longer than 14 days.
